### PR TITLE
Add Redshift/Postgres support for Core Web Vitals

### PR DESCRIPTION
**Pull request by agnessnowplow**
## Description & motivation
Adding Redshift/Postgres support for Core Web Vitals. I decided to add a surrogate_key to measurements as it has always been cumbersome to compare test results without it and it helps in dbt tests as well.

## Checklist
- [x] I have verified that these changes work locally
- [ ] I have updated the README.md (if applicable)
- [x] I have added tests & descriptions to my models (and macros if applicable)
- [ ] I have raised a [documentation](https://github.com/snowplow/documentation) PR if applicable (Link here if required)
- [ ] Is your change a breaking change?

### CODE REVIEW

**Review by rlh1994 — COMMENTED** [review]
Just a few questions really!

File: models/optional_modules/core_web_vitals/scratch/default/snowplow_web_vital_events_this_run.sql [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+  and event_name = 'web_vitals'
+
+  and page_view_id is not null
+
+  -- exclude bot traffic
+
+  {% if var('snowplow__enable_iab', false) %}
+    and not e.iab_spider_or_robot = True
+  {% endif %}
```
[the remark this answers is not in the document]
Comment (agnessnowplow) [reply models/optional_modules/core_web_vitals/scratch/default/snowplow_web_vital_events_this_run.sql:54]: I think originally yes, but now that it is wrapped around the enable variable it doesn't matter, should prob simplify

File: models/optional_modules/core_web_vitals/default/snowplow_web_vital_measurements.sql [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+    {{ grouping_function }}(time_period, device_class) as id_period_and_device,
+    {{ grouping_function }}(geo_country) as id_country,
+    {{ grouping_function }}(geo_country, device_class) as id_country_and_device,
+    percentile_cont(0.{{ var('snowplow__cwv_percentile') }}) within group (order by inp) as inp_{{ var('snowplow__cwv_percentile') }}p
+
+  from prep
+
+  group by grouping sets ((), (page_url, device_class), (device_class), (time_period), (time_period, device_class), (geo_country), (geo_country, device_class))
+
+)
```
[the remark this answers is not in the document]
Comment (rlh1994) [reply models/optional_modules/core_web_vitals/default/snowplow_web_vital_measurements.sql:146]: Classic redshift

File: models/optional_modules/core_web_vitals/databricks/snowplow_web_vital_measurements.sql [hunk 3]
```diff
@@ -31,6 +30,7 @@ with measurements as (
 , measurement_type as (
 
   select
+    {{ dbt_utils.generate_surrogate_key(['page_url', 'device_class', 'geo_country', 'time_period' ]) }} surrogate_key,
```
[the remark this answers is not in the document]
Comment (agnessnowplow) [reply models/optional_modules/core_web_vitals/databricks/snowplow_web_vital_measurements.sql:33]: Could do, for the benefit you highlighted.

File: models/optional_modules/core_web_vitals/scratch/default/snowplow_web_vital_events_this_run.sql [hunk 4]
```diff
@@ -0,0 +1,105 @@
+{{
+  config(
+    tags=["this_run"],
+    enabled=var("snowplow__enable_cwv", false) and target.type in ('redshift', 'postgres') | as_bool()
+  )
+}}
+
+with {{ snowplow_utils.get_sde_or_context(var('snowplow__atomic_schema', 'atomic') if project_name != 'snowplow_web_integration_tests' else target.schema~'_snplw_web_int_tests', 'unstruct_event_com_snowplowanalytics_snowplow_web_vitals_1' if project_name != 'snowplow_web_integration_tests' else 'snowplow_web_cwv_context_stg', lower_limit, upper_limit, 'cwv') }}
+
```
[the remark this answers is not in the document]
Comment (agnessnowplow) [reply models/optional_modules/core_web_vitals/scratch/default/snowplow_web_vital_events_this_run.sql:9]: I only put it here due to the context being different but you are right, I can just add it as a new variable to remove this from there.

File: dbt_project.yml [hunk 5]
```diff
@@ -78,6 +78,7 @@ vars:
     snowplow__enable_cwv: false
     snowplow__cwv_days_to_measure: 28
     snowplow__cwv_percentile: 75
+    snowplow__cwv_context: "{{ source('atomic','unstruct_event_com_snowplowanalytics_snowplow_web_vitals_1') }}"
```
Comment (rlh1994) [inline dbt_project.yml:81]: I think this doesn't work with the get_sde_or_context macro, it needs to just be the raw table name (which is also what I need to do with the base macro part anyway with the rest of them)

File: integration_tests/dbt_project.yml [hunk 6]
```diff
@@ -46,6 +46,8 @@ vars:
     # snowplow__atomic_schema: "{{target.schema ~ '_snplw_web_int_tests'}}"
     # snowplow__events_table: 'snowplow_web_events_stg'
     snowplow__events: "{{ ref('snowplow_web_events_stg') }}"
+    snowplow__atomic_schema: "{{ target.schema }}_snplw_web_int_tests"
```
Comment (rlh1994) [inline integration_tests/dbt_project.yml:49]: Can you delete line 46 in this case? I needed it for the base macro stuff but it's the same as you've done now.

**Review by rlh1994 — CHANGES_REQUESTED** [review]
Just need to change the default variable value then think this is good to go!

**Review by rlh1994 — APPROVED** [review]
List of things this looks:
- Good
List of people it looks that to:
- Me

**Pull request merged**